Interpolated strings (#390)
dotnet/csharpstandard: standard/expressions.md

A *primary_expression* that consists of a *literal* ([§7.4.5](lexical-structure.md#745-literals)) is classified as a value.

### §interpolated-string-expressions Interpolated string expressions

> An *interpolated_string_expression* consists of a `$` character immediately followed by text within `"` characters. Within the quoted text there are zero or more **_interpolations_** delimited by `{` and `}` characters, each of which encloses an *expression* and optional formatting specifications.

Reviewer: Q for @jskeet 

Will the markdown converter parse the `**_` for bold italic, or should we use `***`?

Reviewer 2: Good question - no idea; I believe `***` would work though and is what we've done elsewhere.

Interpolated string expressions have two forms; regular (*interpolated_regular_string_expression*)
and verbatim (*interpolated_verbatim_string_expression*); which are lexically similar to, but differ semantically from, the two forms of string

> literals ([§7.4.5.6](lexical-structure.md#7456-string-literals)).

Reviewer: Note for @Nigel-Ecma 

You can just write the section number for the link, and the tool will fill in the correct section:

```suggestion
literals (§7.4.5.6).
```

Author: OK, changed. Does it make any difference as the long hand version will be there next time, or is it just to help reduce errors?

Reviewer: It won't make any difference later. It's to save you time when you add links.

```ANTLR
interpolated_string_expression
    : interpolated_regular_string_expression
    | interpolated_verbatim_string_expression
    ;

// interpolated regular string expressions

interpolated_regular_string_expression
    : Interpolated_Regular_String_Start Interpolated_Regular_String_Mid?
      ('{' regular_interpolation '}' Interpolated_Regular_String_Mid?)* Interpolated_Regular_String_End
    ;

regular_interpolation
    : expression (',' interpolation_minimum_width)? Regular_Interpolation_Format?
    ;

interpolation_minimum_width
    : constant_expression
    ;

> Interpolated_Regular_String_Start

Reviewer 2: I can't remember the different casing rules we've got, and I probably couldn't check whether these are appropriate anyway. I strongly suspect they're all fine, but I bring it up so that someone with more of a clue can review that aspect specifically.

Author: I suspect it looks a bit off as this is the first place where we really intermingle lexer & parser rules a lot. The casing is correct or ANTLR would have barfed.

    ;

Interpolated_Regular_String_End
    : '"'
    ;

fragment Interpolated_Regular_String_Element
    : Interpolated_Regular_String_Character
    | Simple_Escape_Sequence
    | Hexadecimal_Escape_Sequence
    | Unicode_Escape_Sequence
    | Open_Brace_Escape_Sequence
    | Close_Brace_Escape_Sequence
    ;

fragment Interpolated_Regular_String_Character
    // Any character except " (U+0022), \\ (U+005C), { (U+007B), } (U+007D), and New_Line_Character
    : ~["\\{}\u000D\u000A\u0085\u2028\u2029]
    ;

// interpolated verbatim string expressions

interpolated_verbatim_string_expression
    : Interpolated_Verbatim_String_Start Interpolated_Verbatim_String_Mid?
      ('{' verbatim_interpolation '}' Interpolated_Verbatim_String_Mid?)* Interpolated_Verbatim_String_End
    ;

verbatim_interpolation
    : expression (',' interpolation_minimum_width)? Verbatim_Interpolation_Format?
    ;

Interpolated_Verbatim_String_Start
    : '$@"'
    ;

// the following three lexical rules are context sensitive, see details below
Interpolated_Verbatim_String_Mid
    : Interpolated_Verbatim_String_Element+
    ;

Verbatim_Interpolation_Format
    : ':' Interpolated_Verbatim_String_Element+
    ;

> Interpolated_Verbatim_String_End

Reviewer 2: Do we particularly need this to be separate from Interpolated_Regular_String_End? Could we just have Interpolated_String_End and use it in both rules? Or is that harder due to the context-sensitive nature?

Same applies for some other rules as well, I believe - it's just that this was the simplest one to validate.

Author: Correct, it is the context-sensitivity.

In the working ANTLR grammar there are two (in addition to the standard one) modes, one for each of regular & verbatim interpolated strings. In these two modes some rules are different while others are the same but with different names (as there is a single scope for names). This is a common situation when using modes, nothing special for the use case here.

Reviewer 2: SGTM.

    : '"'
    ;

fragment Interpolated_Verbatim_String_Element
    : Interpolated_Verbatim_String_Character
    | Quote_Escape_Sequence
    | Open_Brace_Escape_Sequence
    | Close_Brace_Escape_Sequence
    ;

fragment Interpolated_Verbatim_String_Character
    // Any character except " (U+0022), { (U+007B) and } (U+007D)
    : ~["{}]
    ;

// lexical fragments used by both regular and verbatim interpolated strings

fragment Open_Brace_Escape_Sequence
    : '{{'
    ;

fragment Close_Brace_Escape_Sequence
    : '}}'
    ;
```

Six of the lexical rules defined above are *context sensitive* as follows:

| Rule | Contextual Requirements |
| :--- | :---------------------- |
| *Interpolated_Regular_String_Mid* | Only recognised after an *Interpolated_Regular_String_Start*, between any following interpolations, and before the corresponding *Interpolated_Regular_String_End*. |

> | *Regular_Interpolation_Format* | Only recognised immediately before the closing delimiter (`}`) of a *regular_interpolation*. |

Reviewer: Should this key on the `:` in the interpolation rather than "Only recognized immediately before the closing `}` delimiter?  Would this rule mean that a conditional (`? :`) expression should parse correctly without parentheses if there's an additional format specifier after the expression?

Author: The tricky part here is the format can contain just about anything, including parentheses etc. So the rule matches if there is a colon, some text, and then a following '}' (which is not part of the token itself) *and* the whole token is at the outmost level.

What we're not making clear here, because its an implementation thing, is that in order to implement the context sensitivity you need to track *all* bracket kinds.

The question is whether the text here is a correct/sufficient description so that implementors can produce a correct implementation. I'll elaborate on the TG2 list…

Reviewer: I like the alternative you suggested to the committee.  Adding it here for discussion:

```suggestion
| *Regular_Interpolation_Format* | Only recognised within a *regular_interpolation*, when the starting colon (`:`) is not nested within any kind of bracket (parentheses/braces/square), and when the following brace (`}`) is the terminating delimiter of a *regular_interpolation*. |
```

Reviewer 2: Agreed with the suggestion without `, and when the following brace (`}`) is the terminating delimiter of a *regular_interpolation*`.

| *Interpolated_Regular_String_End* | Only recognised after an *Interpolated_Regular_String_Start* and only if any intervening tokens are either *Interpolated_Regular_String_Mid*s or tokens that can be part of *regular_interpolation*s, including tokens for any *interpolated_regular_string_expression*s contained within such interpolations. |
| *Interpolated_Verbatim_String_Mid* *Verbatim_Interpolation_Format* *Interpolated_Verbatim_String_End* | Recognition of these three rules follows that of the corresponding rules above with each mentioned *regular* grammar rule replaced by the corresponding *verbatim* one. |

*Note:* The above rules are context sensitive as their definitions overlap with those of

> other tokens in the language. *end note*

Reviewer: All our other "Note"s have been indented:

```suggestion
> *Note:* The above rules are context sensitive as their definitions overlap with those of other tokens in the language. *end note*
```

Author: Fixed

*Note:* The above grammar is not ANTLR-ready due to the context sensitive lexical rules. As with

> other lexer generators ANTLR supports context sensitive lexical rule, for example using its *lexical modes*,

Reviewer 2: lexical rule => lexical rules

(unless I'm misreading)

Author: Thanks, will be fixed.

> but this is an implementation detail and therefore not part of this Standard. *end note*

Reviewer: Same comment (and made the change Jon suggested):

```suggestion
> *Note:* The above grammar is not ANTLR-ready due to the context sensitive lexical rules. As with other lexer generators ANTLR supports context sensitive lexical rules, for example using its *lexical modes*, but this is an implementation detail and therefore not part of this Standard. *end note*
```

Author: Fixed

An *interpolated_string_expression* is classified as a value. If it is immediately converted to `System.IFormattable` or `System.FormattableString` with an implicit interpolated string conversion (§implicit-interpolated-string-conversions), the interpolated string expression has that type. Otherwise, it has the type `string`.

> *Note:* The differences between the possible types an *interpolated_string_expression* may be determined from the documentation for `System.String` ([§C.2](standard-library.md#c2-standard-library-types-defined-in-isoiec-23271)) and `System.FormattableString` ([§C.3](standard-library.md#c3-standard-library-types-not-defined-in-isoiec-23271)). *end note*

Reviewer: More a note for Nigel:

If you just write "§C.2" and "§C.3", the section numbering tool will convert that text to the correct link.

Reviewer: Note formatting:

```suggestion
> *Note:* The differences between the possible types an *interpolated_string_expression* may be determined from the documentation for `System.String` ([§C.2](standard-library.md#c2-standard-library-types-defined-in-isoiec-23271)) and `System.FormattableString` ([§C.3](standard-library.md#c3-standard-library-types-not-defined-in-isoiec-23271)). *end note*
```

Author: Fixed

> The meaning on an interpolation, both *regular_interpolation* and *verbatim_interpolation*, is to format the value of the *expression* as a `string` either according to the format specified by the *Regular_Interpolation_Format* or *Verbatim_Interpolation_Format*, or according to a default format for the type of *expression*. The formatted string is then modified by the *interpolation_minimum_width*, if any, to produce the final `string` to be interpolated into the *interpolated_string_expression*.

Reviewer 2: meaning on => meaning of

Author: Thanks, will be fixed.

> *Note:* How the default format for a type is determined is detailed in the documentation for `System.String` ([§C.2](standard-library.md#c2-standard-library-types-defined-in-isoiec-23271)) and `System.FormattableString` ([§C.3](standard-library.md#c3-standard-library-types-not-defined-in-isoiec-23271)). Descriptions of standard formats, which are identical for *Regular_Interpolation_Format* and *Verbatim_Interpolation_Format*, may be found in the documentation for `System.IFormattable` ([§C.4](standard-library.md#c4-format-specifications)) and in other types in the standard library ([§C](standard-library.md#annex-c-standard-library)). *end note*

Reviewer: same:

```suggestion
> *Note:* How the default format for a type is determined is detailed in the documentation for `System.String` ([§C.2](standard-library.md#c2-standard-library-types-defined-in-isoiec-23271)) and `System.FormattableString` ([§C.3](standard-library.md#c3-standard-library-types-not-defined-in-isoiec-23271)). Descriptions of standard formats, which are identical for *Regular_Interpolation_Format* and *Verbatim_Interpolation_Format*, may be found in the documentation for `System.IFormattable` ([§C.4](standard-library.md#c4-format-specifications)) and in other types in the standard library ([§C](standard-library.md#annex-c-standard-library)). *end note*
```

Author: Fixed

> In an *interpolation_minimum_width* the *constant_expression* shall have an implicit conversion to `int` and it termed the *alignment*, the *field width* is the absolute value of the alignment:

Reviewer 2: it termed => is termed

I'd suggest a semi-colon before "the *field width*" rather than a comma... or possibly even a period, with a new sentence at the end of the paragraph of something like: "The alignment and field width are used to determine the padding, if any, to be used when formatting the string:"

Author: Will fix the typo.

The prose here is not good. The source material often uses "alignment" when they mean "field width" and generally mixes the two up. I've partially fixed this but it is still wrong – the alignment is the *sign* of the *constant_expression* (not its value), the field width is its absolute value. More wordsmithing required...

Reviewer: Proposal: If the only remaining issue is with some of the prose, let's merge this and open a new issue to update the prose in this section.

Reviewer 2: Nigel will tidy it up a little, but not worry too much about achieving perfection: we can adjust slightly later.

The subsequent arguments are the *expression*s from the interpolations, if any, in order.

When an *interpolated_string_expression* contains multiple interpolations, the expressions in those interpolations are evaluated in textual order from the left to right.

*Example*:

This example uses the following format specification features:

 - the `X` format specification which formats integers as uppercase hexadecimal,
 - the default format for a `string` value is the value itself,
 - positive alignment values that right-justify within the specified field width, and
 - that `{{` and `}}` are formatted as `{` and `}` respectively.

Given:

```csharp
string text = "red";
int number = 14;
```

Then:

| Interpolated String Expression       | Equivalent Meaning As `string`                                | Value        |
| :----------------------------------- | :------------------------------------------------------------ | :----------- |
| `$"{text}"`                          | `string.Format("{0}", text)`                                  | `"red"`      |
| `$"{{text}}"`                        | `string.Format("{{text}})`                                    | `"{text}"`   |

> | `$"{ text , 4 }"`                    | `string.Format("{0,4}", text)`                                | `" red"`     |

Reviewer 2: Is it worth having an example that shows the interpolation_minimum_width being an expression using a named constant rather than just digits?

Reviewer 2: Example of valid code that surprised me:

```csharp
const int Width = 4;
string text = "abc";
string formatted = $"{text, Width}";
```

literals (§7.4.5.6).

```ANTLR
interpolated_string_expression
    : interpolated_regular_string_expression
    | interpolated_verbatim_string_expression
    ;

// interpolated regular string expressions

interpolated_regular_string_expression
    : Interpolated_Regular_String_Start Interpolated_Regular_String_Mid?
      ('{' regular_interpolation '}' Interpolated_Regular_String_Mid?)* Interpolated_Regular_String_End
    ;

regular_interpolation
    : expression (',' interpolation_minimum_width)? Regular_Interpolation_Format?
    ;

interpolation_minimum_width
    : constant_expression
    ;

Interpolated_Regular_String_Start
    : '$"'
    ;

// the following three lexical rules are context sensitive, see details below

Interpolated_Regular_String_Mid
    : Interpolated_Regular_String_Element+
    ;

Regular_Interpolation_Format
    : ':' Interpolated_Regular_String_Element+
    ;

Interpolated_Regular_String_End
    : '"'
    ;

fragment Interpolated_Regular_String_Element
    : Interpolated_Regular_String_Character
    | Simple_Escape_Sequence
    | Hexadecimal_Escape_Sequence
    | Unicode_Escape_Sequence
    | Open_Brace_Escape_Sequence
    | Close_Brace_Escape_Sequence
    ;

> fragment Interpolated_Regular_String_Character

Reviewer 2: This didn't come out well in the Word converter :(

![Screenshot (15)](https://user-images.githubusercontent.com/17011/142574873-28effd54-1f6c-4b8a-af96-24547e414371.png)

I'll think about options here. Given that it's only a single rule, how would you feel about putting the comments before the production?
